aws/src/main/java/org/apache/iceberg/aws/glue/DynamoLockManager.java

```diff
@@ -0,0 +1,312 @@
+    }
+  }
+
+  @VisibleForTesting
+  void acquireOnce(String entityId, String ownerId) {
+    GetItemResponse response = dynamo.getItem(GetItemRequest.builder()
+        .tableName(lockTableName)
+        .key(toKey(entityId))
+        .build());
+
+    if (!response.hasItem()) {
+      dynamo.putItem(PutItemRequest.builder()
+          .tableName(lockTableName)
+          .item(toNewItem(entityId, ownerId))
+          .conditionExpression(CONDITION_LOCK_ENTITY_NOT_EXIST)
+          .build());
+    } else {
+      Map<String, AttributeValue> currentItem = response.item();
+
+      try {
+        Thread.sleep(Long.parseLong(currentItem.get(COL_LEASE_DURATION_MS).n()));
+      } catch (InterruptedException e) {
+        throw new IllegalStateException(
+            String.format("Fail to acquire lock %s by %s, interrupted during sleep", entityId, ownerId), e);
+      }
+
+      Map<String, AttributeValue> expressionValues = Maps.newHashMap();
+      expressionValues.put(":eid", AttributeValue.builder().s(entityId).build());
+      expressionValues.put(":vid", AttributeValue.builder().s(currentItem.get(COL_VERSION).s()).build());
+      dynamo.putItem(PutItemRequest.builder()
+          .tableName(lockTableName)
+          .item(toNewItem(entityId, ownerId))
+          .conditionExpression(CONDITION_LOCK_ENTITY_NOT_EXIST_OR_VERSION_MATCH)
+          .expressionAttributeValues(expressionValues)
+          .build());
+    }
+
+    startNewHeartbeat(entityId, ownerId);
+  }
+
+  private void startNewHeartbeat(String entityId, String ownerId) {
+    if (heartbeats.containsKey(entityId)) {
+      heartbeats.remove(entityId).cancel(false);
+    }
+
+    heartbeats.put(entityId, scheduler().scheduleAtFixedRate(() -> dynamo.putItem(PutItemRequest.builder()
+        .tableName(lockTableName)
+        .item(toNewItem(entityId, ownerId))
+        .conditionExpression(CONDITION_LOCK_ID_MATCH)
+        .expressionAttributeValues(toLockIdValues(entityId, ownerId))
+        .build()), 0, heartbeatIntervalMs(), TimeUnit.MILLISECONDS));
+  }
+
+  @Override
+  public boolean release(String entityId, String ownerId) {
+    ScheduledFuture<?> heartbeat = heartbeats.get(entityId);
+    try {
+      dynamo.deleteItem(DeleteItemRequest.builder()
+          .tableName(lockTableName)
+          .key(toKey(entityId))
+          .conditionExpression(CONDITION_LOCK_ID_MATCH)
+          .expressionAttributeValues(toLockIdValues(entityId, ownerId))
+          .build());
+    } catch (ConditionalCheckFailedException e) {
+      LOG.error("Fail to release lock {} by {}, lock entity does not exist or owner not match", entityId, ownerId, e);
+      return false;
+    }
+
+    if (heartbeat != null) {
+      heartbeat.cancel(false);
+    }
+
+    return true;
+  }
+
+  private Map<String, AttributeValue> toKey(String entityId) {
+    Map<String, AttributeValue> key = Maps.newHashMap();
+    key.put(COL_LOCK_ENTITY_ID, AttributeValue.builder().s(entityId).build());
+    return key;
```

Comment:
What about `ImmutableMap.of(COL_LOCK_ENTITY_ID, ...)`? This is probably a nit, but I tend to prefer immutable classes if possible, especially for small maps where a hash lookup might be avoided because there's just one element.

Reply:
FIxed, also updated other similar usages.